api/views/privaterelay.py

```diff
@@ -314,84 +316,134 @@ def terms_accepted_user(request):
     if token == "":
         raise ParseError("Missing FXA Token after 'Bearer'.")
 
-    try:
-        fxa_uid = get_fxa_uid_from_oauth_token(token, use_cache=False)
-    except AuthenticationFailed as e:
-        # AuthenticationFailed exception returns 403 instead of 401 because we are not
-        # using the proper config that comes with the authentication_classes. See:
-        # https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication
-        if isinstance(e.detail, ErrorDetail):
-            return Response(data={"detail": e.detail.title()}, status=e.status_code)
-        else:
-            return Response(data={"detail": e.get_full_details()}, status=e.status_code)
-    status_code = 201
+    user = request.user
+    introspect_response = request.auth
+    if not isinstance(introspect_response, IntrospectionResponse):
+        raise ValueError(
+            "Expected request.auth to be IntrospectionResponse,"
+            f" got {type(IntrospectionResponse)}"
+        )
+    fxa_uid = introspect_response.fxa_id
+
+    existing_sa = False
+    action: str | None = None
+    if isinstance(user, User):
+        socialaccount = SocialAccount.objects.get(user=user, provider="fxa")
+        existing_sa = True
+        action = "found_existing"
+
+    if not existing_sa:
+        # Get the user's profile
+        fxa_profile, response = _get_fxa_profile_from_bearer_token(token)
+        if response:
+            return response
+
+        # Since this takes time, see if another request created the SocialAccount
```

Comment:
question (non-blocking): do you suspect this might be the cause of the `SocialAccount.DoesNotExist` errors? That separate requests are creating/retrieving the `SocialAccount` object?